### PR TITLE
Alexkalanis/refactoring for php7 oop

### PR DESCRIPTION
**Pull request by alex-kalanis**
Update nearly the complete code, possible to use as v2, minimal php7

- added service to access Nearby endpoint (my inital reason)
- enabled extending of some parts (extended service factory, other query libraries)
- added tests (phpunit, phpstan, codestyle)
- added comments and property descriptions

### CODE REVIEW

**Comment by dereuromark** [comment]
Wow, this looks super useful
I love that you also added tests

Would you be interested in re-added this to the maintained fork @ https://github.com/php-collective/google-maps-services ?
- CI
- tests
- new functionality

Note:
- CS is already added, so this can be skipped.
- We can rename the namespace from yidas\googleMaps to `PhpCollective\GoogleMaps`

**Comment by yidas** [comment]
There are two issues I'd like to discuss:
1. Modifying the mininal PHP version to 7.4 is embarrassing, I suggest to keep supporting 7.0 as the current one.
2. Modifying the namespace from lower to upper camel case is the same as above.

Thanks!

**Comment by dereuromark** [comment]
7.4 is definitly the best min version possible at this point (2024).
Even 8.0 is already EOL - https://www.php.net/supported-versions.php
Whats the reason to support versions from like 4 years ago?

I agree that the namespace change looks like a BC break and as such should not be done without majoring, so probably not in this PR.

**Comment by alex-kalanis** [comment]
7.1 is minimum due type control over null. "?string" etc... Even PSR already have problems with php7 (here http message).    
API in src\Client stays the same.

**Comment by alex-kalanis** [comment]
Also I do not expect that this pull request will rot here for 2 years. I set major version in mine repo as 2 - breaking change is then expected.

**Comment by yidas** [comment]
The idea is to take care of users of the ancient system who want to use new services as possible.

How about to keep the mininal PHP version at 7.1 so far with the same namespace.

**Comment by dereuromark** [comment]
There should be no one anymore: https://stitcher.io/blog/php-version-stats-january-2024
Ancient is basically 7.4 (LTS for 7.x)

**Comment by yidas** [comment]
I also referred to that article which revealed the number of 7.1 releases until January 2024.

from my perspective, the priority we can work on right now is to check which code changes are using the 7.4 version restrictions.

**Comment by alex-kalanis** [comment]
Namespaces rewritten.

File: composer.json [hunk 1]
```diff
@@ -10,9 +10,19 @@
         "source": "https://github.com/yidas/google-maps-services-php"
     },
     "require": {
-        "php": ">=5.4",
+        "ext-json": "*",
+        "php": ">=7.4",
```
Comment (dereuromark) [inline composer.json]: 7.1 right?

File: src/Client.php [hunk 2]
```diff
@@ -16,231 +15,68 @@
  * @method array elevation(string $locations, array $params=[])
  * @method array geocode(string $address, array $params=[])
  * @method array reverseGeocode(string $latlng, array $params=[])
- * @method array computeRoutes(array $origin, array $destination, array $body=[], array $fieldMask=[])
+ * @method array computeRoutes(array $origin, array $destination, array $body=[], array $headers=[], array $params=[])
```
Comment (dereuromark) [inline src/Client.php:18]: That looks wrong

File: .github/workflows/code_checks.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  push:
+    branches:
+      - master
+
+jobs:
+  tests:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        php: [ '7.4', '8.0', '8.1' ]
```
Comment (dereuromark) [inline .github/workflows/code_checks.yml]: Usually min max suffices
7.1, 8.3 then

File: tests/ServiceTests/RoutesTest.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+{
+    /**
+     * @throws ServiceException
+     */
+    public function testService(): void
+    {
+        $lib = new Services\Routes(new ApiAuth('test'));
+        $this->assertEquals('https://routes.googleapis.com/directions/v2:computeRoutes', $lib->getPath());
+        $this->assertEquals('POST', $lib->getMethod());
+        $this->assertEquals(null, $lib->getBody());
```
Comment (dereuromark) [inline tests/ServiceTests/RoutesTest.php]: assertNull() usully

Small nitpick: also assertSame() with scalar values to make sure the type also mappes IMO

File: src/Services/Routes.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+     *
+     * @param array<mixed>|null $origin
+     * @param array<mixed>|null $destination
+     * @param array<mixed> $body Full body
+     * @param array<string, string> $headers
+     * @param array<string, string|int|float> $params Query parameters
+     * @throws ServiceException
+     * @return array<string, string|int|float>
+     */
+    public function computeRoutes($origin, $destination, $body=[], array $headers=[], array $params=[]): array
```
Comment (dereuromark) [inline src/Services/Routes.php:59]: This looks different than upstream. I think headers are not used anymore afaik, instead fieldmask
See https://github.com/yidas/google-maps-services-php/pull/15/files

**Comment by dereuromark** [comment]
@yidas I recommend changing the github setting so only "new github contributors" require a manual workflow run init.
This way, the tests will run right away for PRs.

**Comment by dereuromark** [comment]
We should probably wrap this up :)

File: src/Client.php (hunk 2 above)
Comment (alex-kalanis) [reply src/Client.php:18]: \yidas\googleMaps\Services\Routes::computeRoutes - src/Services/Routes.php:59

File: .github/workflows/code_checks.yml (hunk 3 above)
Comment (alex-kalanis) [reply .github/workflows/code_checks.yml]: Adding.

File: src/Client.php (hunk 2 above)
Comment (dereuromark) [reply src/Client.php:18]: https://github.com/yidas/google-maps-services-php/blob/a6300c722d7516980c882352fdccca9497a1700b/src/Routes.php#L31
My point is that there have been changes in upstream to provide fieldmask directly which seems reverted now in your PR.

File: .github/workflows/code_checks.yml (hunk 3 above)
Comment (dereuromark) [reply .github/workflows/code_checks.yml]: My point is that only 7.1 and 8.3 needs testing

**Comment by dereuromark** [comment]
@yidas you still would want to change the github action run setting as per my suggestion

**Comment by alex-kalanis** [comment]
Nearly two years later after initial... [FORK](https://github.com/alex-kalanis/google-maps-php-services) and close. Thanks for your ideas and opinions.

Thanks to @yidas for basics to play with.
Thanks to @dereuromark to idea about Client-independent implementation which makes a huge difference inside the code and other nitpicks.

**Pull request closed without merging**